E2E: Fix artifacts handling in CI

WunderBart opened the pull request:
## What?

Resolves #58995.
Supersedes #57183.

Since the [artifacts became immutable in v4 of actions/upload-artifact](https://github.com/actions/upload-artifact#:~:text=The%20contents%20of%20an%20Artifact%20are%20uploaded%20together%20into%20an%20immutable%20archive.%20They%20cannot%20be%20altered%20by%20subsequent%20jobs%20unless%20the%20Artifacts%20are%20deleted%20and%20recreated%20(where%20they%20will%20have%20a%20new%20ID).%20Both%20of%20these%20factors%20help%20reduce%20the%20possibility%20of%20accidentally%20corrupting%20Artifact%20files.), we must create a separate artifact for each shard (both failures and flaky reports), or we'll get a conflict error (see the attached issue #58995). This is the main issue this PR addresses.

#### Enhancements

For convenience, this PR adds a step to merge artifacts into a single zip so we don't need to remember the shard number. The partial artifacts will be deleted after merging to reduce noise. In case of a rerun, the new artifacts (if any) will be merged with those from the previous run(s) so we don't lose any information. Lastly, unlike currently, the `Report to GitHub` job will only run if a flaky test report zip is produced.

## Testing Instructions

Since this is not manually testable, I've pushed a [tmp commit](https://github.com/WordPress/gutenberg/pull/61338/commits/5e27003bab65914d6ecff335f96fbc199b8ab4f2) that causes two tests to be flaky. Those tests are in two shards (no. 1 & 2) to verify that the artifact upload, merging, and flaky test reporting work as expected. Download the artifacts and confirm the content matches what's reported in CI. Check out the CI output and ensure everything looks as expected. Confirm the [flaky report comment](https://github.com/WordPress/gutenberg/pull/61338#issuecomment-2094346787) contains correct test reports.

_The tmp commit will be reverted when this PR is approved. The reopened flaky report issues will also be closed._

WunderBart committed:
Fix artifacts handling in CI

WunderBart committed:
tmp: fake flaky tests to verify artifacts merging

Mamaduka commented on .github/workflows/end2end-test.yml:
```diff
@@ -53,24 +53,51 @@ jobs:
             - name: Archive debug artifacts (screenshots, traces)
               uses: actions/upload-artifact@65462800fd760344b1a7b4382951275a0abb4808 # v4.3.3
-              if: always()
+              if: ${{ !cancelled() }}
```
What's the difference here?

WunderBart replied:
`always()` causes the step to be always executed, even when the job is cancelled, which I don't think we need. Using `!cancelled()` will skip it if the job is cancelled.

Mamaduka replied:
Thanks for the clarification!

Mamaduka reviewed (approved):
Thank you, @WunderBart!

The generated artifacts and flaky test reports look good to me. Let's not forget to revert the temp commit before merging.

WunderBart committed:
Revert "tmp: fake flaky tests to verify artifacts merging"

This reverts commit 5e27003bab65914d6ecff335f96fbc199b8ab4f2.

WunderBart committed:
Fix boolean output reference
Those values aren't actually booleans, so we need to check by string comparison.

WunderBart commented:
I reverted the tmp commit, and everything seems to work as expected: Artifacts were not produced, so the `Report to Github` job was skipped. Merging!

Mamaduka commented:
@WunderBart, I noticed following errors in e2e test summary - https://github.com/WordPress/gutenberg/actions/runs/9042463717

## Screenshot
![CleanShot 2024-05-11 at 15 19 15](https://github.com/WordPress/gutenberg/assets/240569/df8bef07-6583-48b6-95aa-4e26c61bfa80)

WunderBart commented:
> @WunderBart, I noticed following errors in e2e test summary - https://github.com/WordPress/gutenberg/actions/runs/9042463717
> 
> ## Screenshot
> ![CleanShot 2024-05-11 at 15 19 15](https://private-user-images.githubusercontent.com/240569/329777376-df8bef07-6583-48b6-95aa-4e26c61bfa80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU1OTE3MzgsIm5iZiI6MTcxNTU5MTQzOCwicGF0aCI6Ii8yNDA1NjkvMzI5Nzc3Mzc2LWRmOGJlZjA3LTY1ODMtNDhiNi05NWFhLTRlMjZjNjFiZmE4MC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTEzJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxM1QwOTEwMzhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00MGRmZmM1NWU2YjE1MGI5M2E1OGIzZTMyYTQ3NmZkNWM2MmQxMjNmMGU0YjRlNTQzODg0NmY3YWQ0ODdiNWQwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.A-Q8PIuzUqOmNmYK52MLUr53CIhYAQLT9jk0o5csj7o)

I should have mentioned it in the original PR, but thankfully that's expected. It's because the [merge action](https://github.com/actions/upload-artifact/blob/main/merge/README.md) does not have an [`if-no-files-found` option as the upload artifact has](https://github.com/actions/upload-artifact/#:~:text=%23%20The%20desired,files%2Dfound%3A), so we have to set `continue-on-error` to true, which throws that error. [We used that workaround before as well](https://github.com/WordPress/gutenberg/pull/61338/files#diff-2141f48e1f87cb8f1726e5dcccf0374348734c4afed35a9f4e1cbd70b6310783R111-R112), because the download artifact action also didn't have the _skip if nothing found_ option.

Mamaduka commented:
Thanks for the clarification, @WunderBart!
